mining/mining.go

```diff
@@ -591,56 +602,64 @@ mempoolLoop:
 	// The starting block size is the size of the block header plus the max
 	// possible transaction count size, plus the size of the coinbase
 	// transaction.
-	blockSize := blockHeaderOverhead + uint32(coinbaseTx.MsgTx().SerializeSize())
-	blockSigOps := numCoinbaseSigOps
+	blockWeight := uint32((blockHeaderOverhead * (blockchain.WitnessScaleFactor - 1)) + blockchain.GetTransactionWeight(coinbaseTx))
+	blockSigOpCost := coinbaseSigOpCost
 	totalFees := int64(0)
 
+	// TODO(roasbeef): should be guarded by version bits state check
+	var witnessIncluded bool
+	includeWitness := true
+
 	// Choose which transactions make it into the block.
 	for priorityQueue.Len() > 0 {
 		// Grab the highest priority (or highest fee per kilobyte
 		// depending on the sort order) transaction.
 		prioItem := heap.Pop(priorityQueue).(*txPrioItem)
 		tx := prioItem.tx
 
+		switch {
+		// If segregated witness has not been activated yet, then we
+		// shouldn't include any witness transactions in the block.
+		case tx.MsgTx().HasWitness() && !includeWitness:
```

Comment:
Added a cached version of `HasWitness` to `btcutil.Tx` in my `btcutil` PR that this one depends on.